src/core/lib/security/credentials/ssl/ssl_credentials.cc

```diff
@@ -119,6 +149,42 @@ void grpc_ssl_credentials::set_max_tls_version(
   config_.max_tls_version = max_tls_version;
 }
 
+grpc_security_status grpc_ssl_credentials::initialize_client_handshaker_factory(
+    const grpc_ssl_config* config, const char* pem_root_certs,
+    const tsi_ssl_root_certs_store* root_store,
+    tsi_ssl_session_cache* session_cache) {
+  if (client_handshaker_factory_ != nullptr) {
```

Comment:
I don't think the current strategy works because of the following example:
* User creates SSL credentials object and holds on to it for the lifetime of their application.
* User uses the SSL credentials object to create channel #1 without any session cache in the channel args.
* User uses the SSL credentials object to create channel #2 with a session cache in the channel args.

The user expects that both TLS handshakes are full handshakes, and the cache is populated with a session ticket after the creation of channel #2. However, this will not happen with the current strategy because the handshaker factory is created when we create channel #1 (when there is no session cache), and it is reused when we create channel #2.

To get around examples like this, I'm proposing we slightly tweak the logic as follows:
1. In the `grpc_ssl_credentials` constructor, call `initialize_client_handshaker_factory` with no session cache and store the resulting client handshaker factory as a member field of the `grpc_ssl_credentials` class.
2. In the `grpc_ssl_credentials::create_security_connector` method, try to read the session cache from the channel args, as you're doing now. If the session cache is nullptr, then using the client handshaker factory that is stored as a member field; otherwise, call `initialize_client_handshaker_factory` to build a new client handshaker factory (with the non-null session cache) and we will only use this client handshaker factory for this subchannel.

WDYT?

Reply:
That makes us call `SSL_CTX_new` and make a new `client_handshaker_factory` in the case of a session cache - we lose some efficiency but it would be simpler because then we're keeping this old behavior with a session cache, but if there is no session cache in the channel args then we take the new path tying SSL_CTX to the credential.
Does that make sense? I am good with that approach if so

Reply:
Correct: we maintain the status quo in terms of performance for applications that are specifying the session cache via the channel args, and we land the performance boost (by moving the `SSL_CTX`) for anyone else who is not specifying the session cache via the channel args.